runtime/evaluate.cc

```diff
@@ -509,6 +524,43 @@ class evaluator {
     }
   }
 
+  index_t eval_with_new_context(stmt task) {
+    eval_context context;
+    const let_stmt* closure = as_closure(task);
+    if (closure) task = closure->body;
+    init_context(context, this->context, closure);
+
+    return evaluate(task, context);
+  }
+
+  SLINKY_NO_INLINE index_t eval(const async* op) {
+    index_t task_result = 0;
+    auto task_body = [&]() {
+      task_result = eval_with_new_context(op->task);
+    };
+    ref_count<thread_pool::task> task;
+    thread_pool* pool = context.config->thread_pool;
+    if (pool) {
+      task = pool->enqueue(task_body);
+    } else {
+      task_body();
+    }
+
+    context.reserve(op->sym.id + 1);
+    index_t old_sym = 0;
+    if (op->sym.defined()) context.set(op->sym, reinterpret_cast<index_t>(&*task));
+
+    index_t result = eval_with_new_context(op->body);
```

Comment:
Can't we just evalute body? why does it need a new context?

Reply:
No, we can't change this. `closure` in slinky is not a proper closure, which would actually create a copy of the context and store it in a variable somewhere. In slinky, closures are just a list of all things that are used.

Basically, a more standard closure captures the state for later reading, so once the closure is created, there can't be any more race conditions. But in slinky, closures only prevent writes from being visible, they don't stop reads from racing.